### PR TITLE
Add documentation for key validation

### PR DESCRIPTION
**Pull request by slontis**
This documents the difference between the
EVP_PKEY_XXX_check() calls for the default and fips providers.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: doc/man7/EVP_PKEY-DH.pod [hunk 1]
```diff
@@ -17,15 +17,10 @@ See L<EVP_PKEY-FFC(7)> for more information about FFC keys.
 
 The B<DH> key type uses PKCS#3 format which saves p and g, but not the 'q' value.
 The B<DHX> key type uses X9.42 format which saves the value of 'q' and this
-must be used for FIPS186-4.
+must be used for FIPS186-4. If key validation is required users should be aware
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-DH.pod:20]: Comma after _required_?

File: doc/man7/EVP_PKEY-DH.pod [hunk 2]
```diff
@@ -17,15 +17,10 @@ See L<EVP_PKEY-FFC(7)> for more information about FFC keys.
...
+of the problems associated with FIPS186-4 style parameters as discussed in
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-DH.pod:21]: I don't think _problems_ is a good word here?
What about _nuances_, _details, _specifics_ or _subtleties_ instead?

File: doc/man7/EVP_PKEY-DH.pod [hunk 3]
```diff
@@ -134,6 +129,44 @@ Where s is the security strength of the key which has values of
...
 
+=head2 DH key validation
+
+For B<DHX> that is not a named group the FIPS186-4 standard specifies that the
+values used for FFC parameter generation are also required for parameter
+validation. This means that optional FFC domain parameter values for
+I<seed>, I<pcounter> and I<gindex> or I<hindex> may need to be stored for
+validation purposes.
+For B<DHX> the I<seed> and I<pcounter> can be stored in ASN1 data
+(but the I<gindex> or I<hindex> can not be stored). It is recommended to use a
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-DH.pod:140]: _cannot be so stored_ instead of _can not be stored_.

File: doc/man7/EVP_PKEY-DH.pod [hunk 4]
```diff
@@ -134,6 +129,44 @@ Where s is the security strength of the key which has values of
...
+named safe prime group instead.
+
+For DH keys L<EVP_PKEY_param_check(3)> behaves in the following way:
+The OpenSSL fips provider tests if the parameters are either an approved safe
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-DH.pod:144]: Capitalise FIPS perhaps??

Here and elsewhere, if so.

File: doc/man7/EVP_PKEY-DH.pod [hunk 5]
```diff
@@ -134,6 +129,44 @@ Where s is the security strength of the key which has values of
...
+prime group OR that the FFC parameters conform to FIPS186-4 as defined in
+SP800-56Ar3 'Assurances of Domain-Parameter Validity'.
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-DH.pod:146]: I'd italicise the title instead of quoting it.  Both might be better again.

File: doc/man7/EVP_PKEY-DH.pod [hunk 6]
```diff
@@ -134,6 +129,44 @@ Where s is the security strength of the key which has values of
...
+The OpenSSL default provider uses simpler checks that allows there to be no I<q>
+value for backwards compatability.
+
+For DH keys L<EVP_PKEY_param_check_quick(3)> is equivalent to
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-DH.pod:150]: Consider adding a comma after _For DH keys_ here and below?

File: doc/man7/EVP_PKEY-DH.pod (hunk 1 above)
Comment (paulidale) [inline doc/man7/EVP_PKEY-DH.pod:20]: You've not been consistent with FIPS186-4, you've also written it FIPS-186-4.  Personally, I'd use FIPS 186-4 but consistency is the way to go.

File: doc/man7/EVP_PKEY-DSA.pod [hunk 7]
```diff
@@ -29,6 +29,22 @@ For "fips186_4" this must be either 2048 or 3072.
 For "fips186_2" this must be 1024.
 For "group" this can be any one of 2048, 3072, 4096, 6144 or 8192.
 
+=head2 DSA key validation
+
+For DSA keys L<EVP_PKEY_param_check(3)> behaves in the following way:
+The OpenSSL fips provider conforms to the rules within the FIPS-186-4
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-DSA.pod:35]: See comment about _FIPS186-4_.

File: doc/man7/EVP_PKEY-EC.pod [hunk 8]
```diff
@@ -167,6 +167,26 @@ that m > k3 > k2 > k1 > 0
 
 =back
 
+=head2 EC key validation
+
+For EC keys L<EVP_PKEY_param_check(3)> behaves in the following way:
+For the OpenSSL default provider it tests either
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-EC.pod:173]: Blank line above for consistency.

File: doc/man7/EVP_PKEY-EC.pod [hunk 9]
```diff
@@ -167,6 +167,26 @@ that m > k3 > k2 > k1 > 0
...
+EC_FLAG_CHECK_NAMED_GROUP. The OpenSSL fips provider uses
+L<EC_GROUP_check_named_curve(3)> in order to conform to SP800-56Ar3
+'Assurances of Domain-Parameter Validity'.
+
+For EC keys L<EVP_PKEY_param_check_quick(3)> is equivalent to
+L<EVP_PKEY_param_check(3)>.
+
+For EC keys L<EVP_PKEY_public_check(3)> and L<EVP_PKEY_public_check_quick(3)>
+conform to SP800-56Ar3 'ECC Full Public-Key Validation' and
+'ECC Partial Public-Key Validation'.
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-EC.pod:184]: Add _respectively_ to the end.

File: doc/man7/EVP_PKEY-EC.pod [hunk 10]
```diff
@@ -167,6 +167,26 @@ that m > k3 > k2 > k1 > 0
...
+
+For EC Keys L<EVP_PKEY_private_check(3)> and L<EVP_PKEY_pairwise_check(3)>
+conform to the SP800-56Ar3 'Private key validity' and
+'Owner Assurance of Pair-wise Consistency'.
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-EC.pod:188]: Add _respectively_ to the end.

File: doc/man7/EVP_PKEY-DSA.pod [hunk 11]
```diff
@@ -29,6 +29,22 @@ For "fips186_4" this must be either 2048 or 3072.
...
+default provider uses a much simpler check (see below) for parameter validation,
+unless the seed parameter is set.
+
+For DSA keys L<EVP_PKEY_param_check_quick(3)> behaves in the following way:
+A simple check of L and N and partial g is performed. The default provider
+supports validation of legacy "fips186_2" keys.
+
+For DSA keys L<EVP_PKEY_public_check(3)>, L<EVP_PKEY_private_check(3)> and
+L<EVP_PKEY_pairwise_check(3)> the OpenSSL default and fips providers conform to
+the rules within SP800-56Ar3 for public, private and pairwise tests.
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-DSA.pod:46]: Add _respectively_ to the end.

File: doc/man7/EVP_PKEY-EC.pod (hunk 10 above)
Comment (paulidale) [inline doc/man7/EVP_PKEY-EC.pod:188]: Add _respectively_ to the end.

File: doc/man7/EVP_PKEY-RSA.pod [hunk 12]
```diff
@@ -179,6 +179,27 @@ The auxiliary probable primes.
 
 =back
 
+=head2 RSA key validation
+
+For RSA keys L<EVP_PKEY_param_check(3)> and L<EVP_PKEY_param_check_quick(3)>
+return 1.
```
Comment (paulidale) [inline doc/man7/EVP_PKEY-RSA.pod:185]: Perhaps, add _unconditionally_ to the end if they never fail??

**Review by paulidale — CHANGES_REQUESTED** [review]
Little nits :grin:

Looks good.

File: doc/man7/EVP_PKEY-DH.pod (hunk 1 above)
Comment (slontis) [reply doc/man7/EVP_PKEY-DH.pod:20]: Will use FIPS186-4 as this is consistent with other places (including SP800-)

**Review by paulidale — APPROVED** [review]

File: doc/man7/EVP_PKEY-EC.pod [hunk 13]
```diff
@@ -167,6 +167,26 @@ that m > k3 > k2 > k1 > 0
 
 =back
 
+=head2 EC key validation
+
+For EC keys, L<EVP_PKEY_param_check(3)> behaves in the following way:
+For the OpenSSL default provider it tests either
```
Comment (t8m) [inline doc/man7/EVP_PKEY-EC.pod:173]: Maybe `it uses either` or `it calls either` instead?

File: doc/man7/EVP_PKEY-RSA.pod [hunk 14]
```diff
@@ -179,6 +179,26 @@ The auxiliary probable primes.
...
 
+=head2 RSA key validation
+
+For RSA keys, L<EVP_PKEY_param_check(3)> and L<EVP_PKEY_param_check_quick(3)>
+both return 1 unconditionally.
+
+For RSA keys, L<EVP_PKEY_public_check(3)> conforms to the SP800-56Br1 I<public key
+check> when the OpenSSL FIPS provider is used. The OpenSSL default provider
+performs similiar tests but relaxes the keysize restrictions for backwards
+compatability.
```
Comment (t8m) [inline doc/man7/EVP_PKEY-RSA.pod:190]: Same typo here

**Review by t8m — CHANGES_REQUESTED** [review]
Just some nits

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to both, thanks.

**Pull request closed without merging**